// web/src/pages/Settings.tsx
import React, { useState, useEffect } from 'react'
import { useSupabaseClient, useUser, useSession } from '@supabase/auth-helpers-react'
import { useForm } from 'react-hook-form'
import { Link } from 'react-router-dom'
import toast from 'react-hot-toast'
import subscriptionService, { Subscription } from '../services/subscription-service'

// Types pour le formulaire et abonnements
interface ProfileFormData {
  fullName: string
  company: string
  address: string
  postalCode: string
  city: string
  country: string
  phone: string
  taxNumber: string
  bankName: string
  iban: string
  bic: string
}

const Settings = () => {
  const supabase = useSupabaseClient()
  const user = useUser()
  const session = useSession()
  const [loading, setLoading] = useState(false)
  const [saveSuccess, setSaveSuccess] = useState(false)
  const [subscription, setSubscription] = useState<Subscription | null>(null)
  const [loadingSubscription, setLoadingSubscription] = useState(false)
  const [upgradingPlan, setUpgradingPlan] = useState(false)
  const [cancelingSubscription, setCancelingSubscription] = useState(false)
  
  // React Hook Form
  const { register, handleSubmit, reset, formState: { errors } } = useForm<ProfileFormData>()

  // Chargement du profil utilisateur
  const loadProfile = async () => {
    if (!user) return
    
    setLoading(true)
    try {
      const { data, error } = await supabase
        .from('profiles')
        .select('*')
        .eq('id', user.id)
        .single()
      
      if (error && error.code !== 'PGRST116') {
        throw error
      }
      
      if (data) {
        // Remplissage du formulaire avec les données existantes
        reset({
          fullName: data.full_name || '',
          company: data.company || '',
          address: data.address || '',
          postalCode: data.postal_code || '',
          city: data.city || '',
          country: data.country || '',
          phone: data.phone || '',
          taxNumber: data.tax_number || '',
          bankName: data.bank_name || '',
          iban: data.iban || '',
          bic: data.bic || ''
        })
      }
    } catch (error) {
      console.error('Erreur lors du chargement du profil:', error)
    } finally {
      setLoading(false)
    }
  }

  // Chargement de l'abonnement actuel
  const loadSubscription = async () => {
    if (!session) return
    
    setLoadingSubscription(true)
    try {
      const currentSubscription = await subscriptionService.getCurrentSubscription(session)
      setSubscription(currentSubscription)
    } catch (error) {
      console.error('Erreur lors du chargement de l\'abonnement:', error)
      toast.error('Impossible de charger votre abonnement')
    } finally {
      setLoadingSubscription(false)
    }
  }

  // Initialiser l'upgrade d'un abonnement
  const initiatePlanUpgrade = async (planId: string) => {
    if (!session) return
    
    // Confirmation pour plan Freemium
    if (planId === 'freemium' && !confirm('Voulez-vous activer le plan Freemium ?')) {
      return
    }
    
    const toastId = toast.loading('Préparation du changement de plan...')
    setUpgradingPlan(true)
    
    try {
      // Si c'est un plan freemium, pas besoin de Stripe checkout
      if (planId === 'freemium') {
        await subscriptionService.createCheckoutSession(planId, session)
        toast.success('Plan Freemium activé avec succès!')
        await loadSubscription() // Recharger l'abonnement
      } else {
        // Pour les plans payants, redirection vers Stripe
        const checkoutUrl = await subscriptionService.createCheckoutSession(planId, session)
        toast.dismiss(toastId)
        window.location.href = checkoutUrl
        return // Ne pas exécuter le dismiss après redirection
      }
    } catch (error) {
      console.error('Erreur lors du changement de plan:', error)
      toast.error('Une erreur est survenue. Veuillez réessayer.')
    } finally {
      setUpgradingPlan(false)
      toast.dismiss(toastId)
    }
  }
  
  // Annuler un abonnement
  const cancelSubscription = async () => {
    if (!session || !subscription) return
    
    if (!confirm('Êtes-vous sûr de vouloir annuler votre abonnement ? Cette action est irréversible.')) {
      return
    }
    
    setCancelingSubscription(true)
    const toastId = toast.loading('Annulation de votre abonnement en cours...')
    
    try {
      await subscriptionService.cancelSubscription(session)
      toast.success('Votre abonnement a été annulé')
      await loadSubscription() // Recharger l'abonnement
    } catch (error) {
      console.error('Erreur lors de l\'annulation:', error)
      toast.error('Une erreur est survenue lors de l\'annulation')
    } finally {
      setCancelingSubscription(false)
      toast.dismiss(toastId)
    }
  }
  
  useEffect(() => {
    if (session) {
      loadProfile()
      loadSubscription()
    }
  }, [session])

  // Soumission du formulaire
  const onSubmit = async (data: ProfileFormData) => {
    if (!user) return
    
    setLoading(true)
    setSaveSuccess(false)
    
    try {
      const { error } = await supabase
        .from('profiles')
        .upsert({
          user_id: user.id,
          full_name: data.fullName,
          company: data.company,
          address: data.address,
          postal_code: data.postalCode,
          city: data.city,
          country: data.country,
          phone: data.phone,
          tax_number: data.taxNumber,
          bank_name: data.bankName,
          iban: data.iban,
          bic: data.bic,
          updated_at: new Date().toISOString()
        })
      
      if (error) throw error
      
      setSaveSuccess(true)
      setTimeout(() => setSaveSuccess(false), 3000)
    } catch (error) {
      console.error('Erreur lors de la sauvegarde du profil:', error)
      alert('Une erreur est survenue lors de la sauvegarde. Veuillez réessayer.')
    } finally {
      setLoading(false)
    }
  }

  return (
    <div className="container mx-auto max-w-7xl">
      <div className="md:flex md:items-center md:justify-between mb-6">
        <div className="min-w-0 flex-1">
          <h1 className="text-2xl font-bold leading-7 text-gray-900 sm:text-3xl">
            Paramètres
          </h1>
        </div>
      </div>

      {/* Section Abonnements */}
      <div className="bg-white shadow overflow-hidden sm:rounded-lg mb-8">
        <div className="px-4 py-5 sm:px-6">
          <h2 className="text-lg leading-6 font-medium text-gray-900">
            Abonnement et Facturation
          </h2>
          <p className="mt-1 max-w-2xl text-sm text-gray-500">
            Gérez votre abonnement et consultez votre utilisation.
          </p>
        </div>
        <div className="border-t border-gray-200 px-4 py-5 sm:px-6">
          {loadingSubscription ? (
            <div className="py-6 text-center">
              <div className="animate-spin rounded-full h-10 w-10 border-t-2 border-b-2 border-blue-500 mx-auto"></div>
              <p className="mt-2 text-gray-500">Chargement de votre abonnement...</p>
            </div>
          ) : subscription ? (
            <div>
              <div className="flex flex-col md:flex-row md:justify-between md:items-center">
                <div>
                  <h3 className="text-xl font-semibold text-gray-800">
                    Plan {subscription.plan_type === 'freemium' ? 'Freemium' : 
                          subscription.plan_type === 'standard' ? 'Standard' : 'Premium'}
                  </h3>
                  <p className="text-sm text-gray-500 mt-1">
                    Statut: <span className={`font-semibold ${subscription.status === 'active' ? 'text-green-600' : 
                                              subscription.status === 'canceled' ? 'text-red-600' : 'text-orange-600'}`}>
                      {subscription.status === 'active' ? 'Actif' : 
                       subscription.status === 'canceled' ? 'Annulé' : 
                       subscription.status === 'trialing' ? 'Période d\'essai' : subscription.status}
                    </span>
                  </p>
                </div>
                
                {subscription.plan_type !== 'premium' && (
                  <div className="mt-4 md:mt-0">
                    <button
                      onClick={() => initiatePlanUpgrade(subscription.plan_type === 'freemium' ? 'standard' : 'premium')}
                      disabled={upgradingPlan}
                      className="btn btn-primary"
                    >
                      {upgradingPlan ? 'En cours...' : `Passer au plan ${subscription.plan_type === 'freemium' ? 'Standard' : 'Premium'}`}
                    </button>
                  </div>
                )}
              </div>
              
              <div className="mt-6 grid grid-cols-1 md:grid-cols-3 gap-4">
                <div className="bg-gray-50 p-4 rounded-lg">
                  <h4 className="font-medium text-gray-700">Facturation</h4>
                  <p className="text-sm text-gray-500 mt-1">
                    Prochain renouvellement: {subscription.current_period_end ? 
                      new Date(subscription.current_period_end).toLocaleDateString('fr-FR') : 'N/A'}
                  </p>
                </div>
                
                <div className="bg-gray-50 p-4 rounded-lg">
                  <h4 className="font-medium text-gray-700">Utilisation</h4>
                  <div className="mt-1 flex items-center">
                    <div className="flex-1">
                      <div className="bg-gray-200 rounded-full h-2.5 w-full">
                        <div className={`h-2.5 rounded-full ${
                  subscription.invoice_limit === null ? 'bg-blue-600' : 
                  (subscription.invoice_usage / subscription.invoice_limit) > 0.9 ? 'bg-red-600' : 
                  (subscription.invoice_usage / subscription.invoice_limit) > 0.7 ? 'bg-yellow-500' : 
                  'bg-blue-600'}`} 
                     style={{ width: `${
                       subscription.invoice_limit === null ? 50 : 
                       (subscription.invoice_usage / subscription.invoice_limit) * 100
                     }%` }}>
                        </div>
                      </div>
                    </div>
                    <span className="ml-2 text-sm text-gray-700">
                      {subscription.invoice_usage} / {subscription.invoice_limit || 'Illimité'}
                    </span>
                  </div>
                  <p className="text-xs text-gray-500 mt-1">
                    Factures utilisées ce mois-ci
                  </p>
                </div>
                
                <div className="bg-gray-50 p-4 rounded-lg">
                  <h4 className="font-medium text-gray-700">Historique</h4>
                  <p className="text-sm text-gray-500 mt-1">
                    Abonné depuis: {subscription.start_date ? 
                      new Date(subscription.start_date).toLocaleDateString('fr-FR') : 'N/A'}
                  </p>
                </div>
              </div>
              
              {subscription.plan_type !== 'freemium' && (
                <div className="mt-6 text-right">
                  <button
                    className="text-sm text-red-600 hover:text-red-800"
                    onClick={cancelSubscription}
                    disabled={cancelingSubscription || subscription.status === 'canceled'}
                  >
                    {cancelingSubscription ? 'Annulation en cours...' : 
                     subscription.status === 'canceled' ? 'Abonnement déjà annulé' : 
                     'Annuler l\'abonnement'}
                  </button>
                </div>
              )}
            </div>
          ) : (
            <div className="py-6 text-center">
              <p className="text-gray-600">Vous n'avez pas encore d'abonnement actif.</p>
              <div className="mt-4">
                <button
                  onClick={() => initiatePlanUpgrade('freemium')}
                  className="btn btn-primary"
                  disabled={upgradingPlan}
                >
                  {upgradingPlan ? 'En cours...' : 'Activer le plan Freemium'}
                </button>
              </div>
            </div>
          )}
        </div>
      </div>
      
      {/* Section Informations de facturation */}
      <div className="bg-white shadow overflow-hidden sm:rounded-lg">
        <div className="px-4 py-5 sm:px-6">
          <h2 className="text-lg leading-6 font-medium text-gray-900">
            Informations de facturation
          </h2>
          <p className="mt-1 max-w-2xl text-sm text-gray-500">
            Ces informations apparaîtront sur vos factures.
          </p>
        </div>
        <div className="border-t border-gray-200 px-4 py-5 sm:px-6">
          {loading && !user ? (
            <div className="py-10 text-center">
              <div className="animate-spin rounded-full h-12 w-12 border-t-2 border-b-2 border-blue-500 mx-auto"></div>
              <p className="mt-3 text-gray-500">Chargement des paramètres...</p>
            </div>
          ) : (
            <form onSubmit={handleSubmit(onSubmit)} className="space-y-6">
              {/* Message de succès */}
              {saveSuccess && (
                <div className="rounded-md bg-green-50 p-4">
                  <div className="flex">
                    <div className="flex-shrink-0">
                      <svg className="h-5 w-5 text-green-400" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20" fill="currentColor">
                        <path fillRule="evenodd" d="M10 18a8 8 0 100-16 8 8 0 000 16zm3.707-9.293a1 1 0 00-1.414-1.414L9 10.586 7.707 9.293a1 1 0 00-1.414 1.414l2 2a1 1 0 001.414 0l4-4z" clipRule="evenodd" />
                      </svg>
                    </div>
                    <div className="ml-3">
                      <p className="text-sm font-medium text-green-800">
                        Vos paramètres ont été enregistrés avec succès.
                      </p>
                    </div>
                  </div>
                </div>
              )}

              <div className="grid grid-cols-1 gap-y-6 gap-x-4 sm:grid-cols-6">
                <div className="sm:col-span-3">
                  <label htmlFor="fullName" className="block text-sm font-medium text-gray-700">
                    Nom complet
                  </label>
                  <div className="mt-1">
                    <input
                      type="text"
                      id="fullName"
                      className="input"
                      {...register('fullName', { required: "Le nom est requis" })}
                    />
                    {errors.fullName && (
                      <p className="mt-1 text-sm text-red-600">{errors.fullName.message}</p>
                    )}
                  </div>
                </div>

                <div className="sm:col-span-3">
                  <label htmlFor="company" className="block text-sm font-medium text-gray-700">
                    Entreprise
                  </label>
                  <div className="mt-1">
                    <input
                      type="text"
                      id="company"
                      className="input"
                      {...register('company')}
                    />
                  </div>
                </div>

                <div className="sm:col-span-6">
                  <label htmlFor="address" className="block text-sm font-medium text-gray-700">
                    Adresse
                  </label>
                  <div className="mt-1">
                    <input
                      type="text"
                      id="address"
                      className="input"
                      {...register('address')}
                    />
                  </div>
                </div>

                <div className="sm:col-span-2">
                  <label htmlFor="postalCode" className="block text-sm font-medium text-gray-700">
                    Code postal
                  </label>
                  <div className="mt-1">
                    <input
                      type="text"
                      id="postalCode"
                      className="input"
                      {...register('postalCode')}
                    />
                  </div>
                </div>

                <div className="sm:col-span-2">
                  <label htmlFor="city" className="block text-sm font-medium text-gray-700">
                    Ville
                  </label>
                  <div className="mt-1">
                    <input
                      type="text"
                      id="city"
                      className="input"
                      {...register('city')}
                    />
                  </div>
                </div>

                <div className="sm:col-span-2">
                  <label htmlFor="country" className="block text-sm font-medium text-gray-700">
                    Pays
                  </label>
                  <div className="mt-1">
                    <input
                      type="text"
                      id="country"
                      className="input"
                      {...register('country')}
                    />
                  </div>
                </div>

                <div className="sm:col-span-3">
                  <label htmlFor="phone" className="block text-sm font-medium text-gray-700">
                    Téléphone
                  </label>
                  <div className="mt-1">
                    <input
                      type="text"
                      id="phone"
                      className="input"
                      {...register('phone')}
                    />
                  </div>
                </div>

                <div className="sm:col-span-3">
                  <label htmlFor="taxNumber" className="block text-sm font-medium text-gray-700">
                    Numéro de TVA
                  </label>
                  <div className="mt-1">
                    <input
                      type="text"
                      id="taxNumber"
                      className="input"
                      {...register('taxNumber')}
                    />
                  </div>
                </div>

                <div className="sm:col-span-6">
                  <h3 className="text-lg font-medium text-gray-900">Informations bancaires</h3>
                  <p className="text-sm text-gray-500">
                    Ces informations apparaîtront sur vos factures pour recevoir les paiements.
                  </p>
                </div>

                <div className="sm:col-span-2">
                  <label htmlFor="bankName" className="block text-sm font-medium text-gray-700">
                    Banque
                  </label>
                  <div className="mt-1">
                    <input
                      type="text"
                      id="bankName"
                      className="input"
                      {...register('bankName')}
                    />
                  </div>
                </div>

                <div className="sm:col-span-3">
                  <label htmlFor="iban" className="block text-sm font-medium text-gray-700">
                    IBAN
                  </label>
                  <div className="mt-1">
                    <input
                      type="text"
                      id="iban"
                      className="input"
                      {...register('iban')}
                    />
                  </div>
                </div>

                <div className="sm:col-span-1">
                  <label htmlFor="bic" className="block text-sm font-medium text-gray-700">
                    BIC
                  </label>
                  <div className="mt-1">
                    <input
                      type="text"
                      id="bic"
                      className="input"
                      {...register('bic')}
                    />
                  </div>
                </div>
              </div>

              <div className="flex justify-end">
                <button
                  type="submit"
                  className="btn btn-primary"
                  disabled={loading}
                >
                  {loading ? 'Enregistrement...' : 'Enregistrer'}
                </button>
              </div>
            </form>
          )}
        </div>
      </div>
    </div>
  )
}

// Utilitaires pour l'affichage de l'abonnement
function getUtilizationPercentage(subscription: Subscription): number {
  if (!subscription.invoice_limit) return 0;
  return Math.min(100, Math.round((subscription.invoice_usage / subscription.invoice_limit) * 100));
}

function getUtilizationColorClass(subscription: Subscription): string {
  if (!subscription.invoice_limit) return 'bg-blue-500';
  
  const percentage = getUtilizationPercentage(subscription);
  if (percentage < 70) return 'bg-green-500';
  if (percentage < 90) return 'bg-yellow-500';
  return 'bg-red-500';
}

export default Settings
